app/components/custom_flat_list.js

```diff
@@ -0,0 +1,240 @@
+         */
+        extraData: PropTypes.object,
+
+        /*
+         * A function called when an item in the list is pressed. Receives the item that was pressed as an argument.
+         */
+        onRowPress: PropTypes.func,
+
+        /*
+         * A function called when the end of the list is reached. This can be triggered before this list end is reached
+         * by changing onListEndReachedThreshold.
+         */
+        onListEndReached: PropTypes.func,
+
+        /*
+         * How soon before the end of the list onListEndReached should be called.
+         */
+        onListEndReachedThreshold: PropTypes.number,
+
+        /*
+         * Whether or not to display the loading text.
+         */
+        loading: PropTypes.bool,
+
+        /*
+         * The text displayed when loading is set to true.
+         */
+        loadingText: PropTypes.object,
+
+        /*
+         * How many items to render when the list is first rendered.
+         */
+        initialNumToRender: PropTypes.number,
+    };
+
+    static defaultProps = {
+        showNoResults: true,
+        keyExtractor: (item) => {
+            return item.id;
+        },
+        onListEndReached: emptyFunction,
+        onListEndReachedThreshold: 50,
+        loadingText: null,
+        initialNumToRender: 10,
+    };
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            items: props.items,
+        };
+    }
+
+    componentWillReceiveProps(nextProps) {
+        if (nextProps.items !== this.props.items) {
+            this.setState({
+                items: nextProps.items,
+            });
+        }
+    }
+
+    renderItem = ({item}) => {
+        const props = {
+            id: item.id,
+            item,
+            onPress: this.props.onRowPress,
+        };
+
+        // Allow passing in a component like UserListRow or ChannelListRow
+        if (this.props.renderItem.prototype.isReactElement) {
+            const RowComponent = this.props.renderItem;
+            return <RowComponent {...props}/>;
+        }
+
+        return this.props.renderItem(props);
+    };
+
+    renderItemSeparator = () => {
+        const {theme} = this.props;
+        const style = getStyleFromTheme(theme);
+
+        return <View style={style.separator}/>;
+    };
+
+    renderFooter = () => {
+        const {theme} = this.props;
+        const style = getStyleFromTheme(theme);
+
+        if (!this.props.loading || !this.props.loadingText) {
+            return null;
+        }
+
+        if (this.props.items.length === 0) {
+            return null;
+        }
+
+        return (
+            <View style={style.loading}>
+                <FormattedText
+                    {...this.props.loadingText}
+                    style={style.loadingText}
+                />
+            </View>
+        );
+    };
+
+    renderEmptyList = () => {
+        const {theme} = this.props;
```

Comment:
nothing to do now, but I once #1584 goes in we should use the component for `NoResults` here, also we should use the new `FailedNetworkAction` component for failed requests as well